shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -405,83 +435,261 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
+    }
+  }
+
+  private class CachedBatchIteratorConsumer(
+     cbIter: Iterator[CachedBatch],
+     cacheAttributes: Seq[Attribute],
+     selectedAttributes: Seq[Attribute],
+     sharedHadoopConf: Configuration,
+     sharedConf: SQLConf) {
+
+    val requestedSchema = getCatalystSchema(selectedAttributes, cacheAttributes)
+    val options = HadoopReadOptions.builder(sharedHadoopConf).build()
+
+    def getInternalRowIterator: Iterator[InternalRow] = {
+
+      val cacheSchema = getCatalystSchema(cacheAttributes, cacheAttributes)
+
+      new Iterator[InternalRow]() {
+
+        var iter: Iterator[InternalRow] = null
+
+        override def hasNext: Boolean = {
+          // go over the batch and get the next non-degenerate iterator
+          // and return if it hasNext
+          while ((iter == null || !iter.hasNext) && cbIter.hasNext) {
+            iter = convertCachedBatchToInternalRowIter
+          }
+          iter != null && iter.hasNext
+        }
+
+        override def next(): InternalRow = {
+          // will return the next InternalRow if hasNext() is true, otherwise throw
+          if (hasNext) {
+            iter.next()
+          } else {
+            throw new NoSuchElementException("no elements found")
           }
+        }
 
-          override def next(): InternalRow = {
-            // will return the next InternalRow if hasNext() is true, otherwise throw
-            if (hasNext) {
-               iter.next()
-            } else {
-              throw new NoSuchElementException("no elements found")
+        def convertCachedBatchToInternalRowIter: Iterator[InternalRow] = {
+          val parquetCachedBatch = cbIter.next().asInstanceOf[ParquetCachedBatch]
+          val inputFile = new ByteArrayInputFile(parquetCachedBatch.buffer)
+          val parquetFileReader = ParquetFileReader.open(inputFile, options)
+          val parquetSchema = parquetFileReader.getFooter.getFileMetaData.getSchema
+
+          val unsafeRows = new ArrayBuffer[InternalRow]
+          import org.apache.parquet.io.ColumnIOFactory
+          var pages = parquetFileReader.readNextRowGroup()
+          while (pages != null) {
+            val rows = pages.getRowCount
+            val columnIO = new ColumnIOFactory().getColumnIO(parquetSchema)
+            val recordReader =
+              columnIO.getRecordReader(pages, new ParquetRecordMaterializer(parquetSchema,
+                cacheSchema.toStructType,
+                new ParquetToSparkSchemaConverter(sharedHadoopConf), None /*convertTz*/,
+                LegacyBehaviorPolicy.CORRECTED))
+            for (i <- 0 until rows.toInt) {
+              val row = recordReader.read
+              unsafeRows += row.copy()
             }
+            pages = parquetFileReader.readNextRowGroup()
           }
+          parquetFileReader.close()
+          val iter = unsafeRows.iterator
+          val unsafeProjection =
+            GenerateUnsafeProjection.generate(requestedSchema, cacheSchema)
```

Comment:
This concerns me a little it indicates that we are throwing out rows after parquet read them instead of doing predicate push down.  I am fine with this for now, but we should have a follow on issue to fix it.

Reply:
#1019